### DISCUSSION TITLE
Adding new sensors (like Radar)

### DISCUSSION THREAD

**Discussion by chcardoz** [microsoft/AirSim#3322] [Support Q&A] [answered]
Hello, I am working on a project for my school and we are using airsim to simulate autonomous vehicles. The LiDAR sensor has been very useful. But, I was looking to include a radar sensor as well but there isnt one. I tried looking into the docs to add new sensors, but I don't know where to start in the codebase. I would greatly appreciate some help with:

1) If I wanted to add a custom sensor, where do I start in the codebase? what steps should i follow?
2) How do I expose that custom sensor to the Python client apis so that I can visualize the data?

Any help is greatly appreciated.

**Comment by ahmed-elsaharti** [comment, marked as answer]
Hi @chcardoz , others will probably be able to provide more help with this, but I personally would start with looking at PRs related to what I'm trying to do when I want to change or add anything to Airsim. For example, #1424 (the PR where LiDAR got added) could be a good place to start (in addition to [other PRs related to the topic)](https://github.com/microsoft/AirSim/pulls?q=is%3Apr+lidar).
Same thing with the Python APIs. I'd look at any PR where a new API was added to get how to get things going.

Also, as far as I remember, RADAR data points are similar (in concept) to LiDAR data points. Why not look into just using the data provided by the currently available LiDAR sensor and manipulating it on your client (ie: a python script) to simulate RADAR data?
The reason I'm suggesting this is that I'm assuming it would be far easier to work on the ray-casted LiDAR points from the API than to develop a new sensor.

**Reply by rajat2004** [reply]
Yeah, just adding to this, here's a related issue for Radar - https://github.com/microsoft/AirSim/issues/2863
You can also see PRs https://github.com/microsoft/AirSim/pull/2807 and https://github.com/microsoft/AirSim/pull/2810 which make some modifications to the existing sensors and will help in figuring out the files required to modify

**Reply by chcardoz** [reply]
@ahmed-elsaharti thank you for the reply, you are right that RADAR and LiDAR both measure time of flight and I would have to use ray-casting for both of them. So do you have an idea of what settings I can tweak in the settings.json  file to make it behave like RADAR vs a LiDAR?

**Reply by ahmed-elsaharti** [reply]
@chcardoz if you're trying to simulate a single plane RADAR, I would set `VerticalFOVUpper` and `VerticalFOVLower` to 0 just to get a single plane of points. You could tinker with `RotationsPerSecond` to set how fast the rays rotate and `PointsPerSecond` (which based on my experiments yesterday are equal to the number of points returned by the API call *10, ie: for the scan to contain 10 points, i would set PointsPerSecond to 100). Also, make sure to set `DrawDebugPoints` to true to be able to view the behavior of the beam.
Generally, two ideas that come to mind here are:
1. To set a particular rotation velocity and monitor the first point returned. Since that point is the x,y,z of the surface the beam hit, it'll practically act like a RADAR beam scanning objects around the drone. (you can ofcourse use the x and y coordinates of the returned point to calculate the 'range' and angle as a RADAR sensor would)
2. Set the LiDAR to scan an entire circle of points (ie 360 points atleast per scan) and have your python script just iterate through those points and return just one of them. Not sure if this makes sense, but if it doesn't let me know and i could try to draw a diagram.

**Reply by chcardoz** [reply]
@ahmed-elsaharti , I tweaked it to get 360 points per second in every scan. That would mean in the point cloud, I would get 360 points. What did you mean exactly when you said iterate through the points and return only one?
Here is my `settings.json` for reference: 
```
{
  "SeeDocsAt": "https://github.com/Microsoft/AirSim/blob/master/docs/settings.md",
  "SettingsVersion": 1.2,
  "SimMode": "Car",
  "Vehicles": {
    "Car": {
      "VehicleType": "PhysXCar",
      "AutoCreate": true,
      "Sensors": {
        "LidarSensor1": {
          "SensorType": 6,
          "Enabled": true,
          "NumberOfChannels": 16,
          "Range": 100,
          "RotationsPerSecond": 1,
          "PointsPerSecond": 360,
          "X": 0,
          "Y": 0,
          "Z": -1,
          "Roll": 0,
          "Pitch": 0,
          "Yaw": 0,
          "VerticalFOVUpper": 0,
          "VerticalFOVLower": 0,
          "HorizontalFOVStart": 0,
          "HorizontalFOVEnd": 360,
          "DrawDebugPoints": true,
          "DataFrame": "SensorLocalFrame"
        }
      }
    }
  }
}
```
thank you!

**Reply by ahmed-elsaharti** [reply]
@chcardoz , firstly, in my testing, a PointsPerSecond value of 360 will only produce ~36 points (36 points of x y z so a total of ~108 values in the point_cloud variable returned) per reading, to get 360 points per reading set it to 3600 (I honestly have no idea why this is the case). Also, this might produce (from what i'm seeing through the DebugPoints) a group of 360 scanned points that keep rotating:
![image](https://user-images.githubusercontent.com/49626509/105397348-8fe9c500-5be6-11eb-9425-fe4f906f13f0.png)
in this case i would just use the first point (client.getLidarData().point_cloud[0]) in the returned list. This is because as time progresses the client.getLidarData().point_cloud[0] ray rotates.
![image](https://user-images.githubusercontent.com/49626509/105397968-59607a00-5be7-11eb-9d5e-27fbd460b0e4.png)
So if we only look at client.getLidarData().point_cloud[0], it would be rotating scanning ray just like a RADAR as time progresses:
![image](https://user-images.githubusercontent.com/49626509/105398227-b5c39980-5be7-11eb-98fa-e97933d3ffef.png)

What i meant when i said do a 360 scan and iterate through it is that if you scan an entire circle around the car using the LiDAR sensor in each reading:
![image](https://user-images.githubusercontent.com/49626509/105398350-e1df1a80-5be7-11eb-9e08-97ede73e1ee7.png)
you could start at reading 0 then as time progresses look at the next ray and so on. This would practically only work though if each ray is not rotating (ie ray 0 is always forward and so on). This way you would be simulating the rotation without having the rays rotate. In the example in the previous paragraph you would have been using the sensor's rotation rather than having to simulate it. 

These are essentially two ways of tackling the same problem, they should bothresult in 1 distance/position reading that corresponds to a 'ray' that is rotating around the car.

Also, i think the way the LiDAR sensor works in Airsim is that if the point being scanned (ie the ray being casted) hits beyond the set range of the sensor it doesnt return a point at all (as opposed to returning a '0' like some LiDARs do). Not entirely sure abt this but in my testing the number of points returned did fluctuate based on where my car/drone was

**Reply by chcardoz** [reply]
Wow! thank you so much
